[Feat] 주문 접수 기능 개발

tkdwls4453 opened the pull request:
> PR 당 코드의 변경은 300줄이 넘어가지 않도록 노력하기

## 📝 요약

주문 접수 기능 개발
주문 정보, 쿠폰 정보, 배송 정보를 입렵받아 주문을 접수합니다.

## 💁‍♂️ 이슈

재고 감소, 쿠폰 적용 서비스가 개발되지 않아 가짜 객체로 우선 개발했습니다.

## 🔀 변경사항
* [x] 주문 접수 API 개발
* [x] 테스트 코드 작성
* [x] Money Converter 개발

## 🔗 이슈번호
#40

tkdwls4453 committed:
style: 필요없는 코드 제거

CI 테스트를 위해 했던 필요없는 코드 제거

tkdwls4453 committed:
feat: 주문 접수 도메인 및 컨트롤러 개발 #40

tkdwls4453 committed:
test: 주문 접수 컨트롤러 테스트 코드 작성 #40

tkdwls4453 committed:
refactor: Money 값 객체 사용 및 주문 도메인에 주문 시간 추가 #40

tkdwls4453 committed:
refactor: command 클래스에 from() 함수 추가 #40

request -> command 변경을 돕는 from() 함수를 추가하여 중복 코드 제거

tkdwls4453 committed:
test: 주문 접수 서비스의 뼈대 테스트 코드 작성 #40

tkdwls4453 committed:
feat: 주문 도메인 관련 기능 개발 #40

- 주문 초기화
- 주문 실패
- 주문 진행

tkdwls4453 committed:
test: 주문 도메인 기능 테스트 코드 작성 #40

tkdwls4453 committed:
test: 주문 관련 테스트에 필요한 데이터 추가 #40

tkdwls4453 committed:
test: 도메인에 주문시간 필드 추가하면서 테스트 코드 수정 #40

tkdwls4453 committed:
feat: Money VO 기능 개발

Money VO 는 모든 서비스에서 공용으로 사용될 가능성이 크기 때문에 common 모듈로 뺄 것을 고려

tkdwls4453 committed:
feat: 주문 서비스에서 사용할 예외 추가 #40

tkdwls4453 committed:
style: 쿠폰에서 사용될 타입 주석 추가

추후에 쿠폰 서비스 관리 팀원과 얘기하여 type 을 enum으로 관리할 예정

tkdwls4453 committed:
feat: 주문 생성 기능 개발 #40

tkdwls4453 committed:
test: 주문 생성 기능 테스트 코드 작성 #40

tkdwls4453 committed:
feat: 쿠폰 적용과 재고 감소 가짜 서비스 생성 #40

쿠폰 서비스와 재고 서비스 부분이 개발되지 않아 아무것도 하지 않는 가짜 adpater 를 생성하였다.

tkdwls4453 committed:
feat: Money 값 객체의 conveter 작성 #40

- Money VO 는 공통으로 사용될 가능성이 높아 common 모듈로 이동
- request 의 int -> Money 로 변경해주는 converter 작성
- reponse 의 Money -> int 로 변경해주는 conveter 작성

tkdwls4453 committed:
refactor: 주문에 꼭 필요한 데이터만 남김 #40

- 주문 신발 대신 범용적인 이름인 주문 아이템을 사용
- 주문에 꼭 필요한 데이터만 필드로 남김

tkdwls4453 committed:
refactor: 필드 이름 변경으로 인한 변경점과 필요한 데이터만 넘김 #40

tkdwls4453 committed:
refactor: 필드 이름 변경으로 인한 변경점 수정 #40

tkdwls4453 committed:
feat: 주문 접수 repository 작성 #40

OrderEntity 와 OrderItemEntity 를 연관관계 매핑이 아닌 @ElementCollection 으로 매핑하였음
-> OrderItem의 변경은 Order 로만 가능하게 하여 모든 변경점을 Aggragate Root 에서만 하도록 함

tkdwls4453 committed:
test: 주문 접수 기능의 Persistence Layer 테스트 코드 작성 #40

tkdwls4453 commented:
### 📌 Test Coverage Report
|Overall Project|96.93% **`-3.07%`**|:green_apple:|
|:-|:-|:-:|
|Files changed|96.93%|:green_apple:|
<br>

|File|Coverage||
|:-|:-|:-:|
|[Order.java](https://github.com/tkdwls4453/ecommerce-msa-server/blob/acb089df6cdedeca8419fcb32e851eb5b5fed113/order-service%2Fsrc%2Fmain%2Fjava%2Fcom%2Fmsa%2Forder%2Fdomain%2FOrder.java)|100%|:green_apple:|
|[OrderStatus.java](https://github.com/tkdwls4453/ecommerce-msa-server/blob/acb089df6cdedeca8419fcb32e851eb5b5fed113/order-service%2Fsrc%2Fmain%2Fjava%2Fcom%2Fmsa%2Forder%2Fdomain%2FOrderStatus.java)|100%|:green_apple:|
|[InvalidTotalPriceException.java](https://github.com/tkdwls4453/ecommerce-msa-server/blob/acb089df6cdedeca8419fcb32e851eb5b5fed113/order-service%2Fsrc%2Fmain%2Fjava%2Fcom%2Fmsa%2Forder%2Fexception%2FInvalidTotalPriceException.java)|100%|:green_apple:|
|[InvalidCouponTypeException.java](https://github.com/tkdwls4453/ecommerce-msa-server/blob/acb089df6cdedeca8419fcb32e851eb5b5fed113/order-service%2Fsrc%2Fmain%2Fjava%2Fcom%2Fmsa%2Forder%2Fexception%2FInvalidCouponTypeException.java)|100%|:green_apple:|
|[OrderErrorCode.java](https://github.com/tkdwls4453/ecommerce-msa-server/blob/acb089df6cdedeca8419fcb32e851eb5b5fed113/order-service%2Fsrc%2Fmain%2Fjava%2Fcom%2Fmsa%2Forder%2Fexception%2FOrderErrorCode.java)|100%|:green_apple:|
|[CreateNewOrderCommand.java](https://github.com/tkdwls4453/ecommerce-msa-server/blob/acb089df6cdedeca8419fcb32e851eb5b5fed113/order-service%2Fsrc%2Fmain%2Fjava%2Fcom%2Fmsa%2Forder%2Fapplication%2Fport%2Fin%2FCreateNewOrderCommand.java)|100%|:green_apple:|
|[OrderController.java](https://github.com/tkdwls4453/ecommerce-msa-server/blob/acb089df6cdedeca8419fcb32e851eb5b5fed113/order-service%2Fsrc%2Fmain%2Fjava%2Fcom%2Fmsa%2Forder%2Fadapter%2Fin%2Fweb%2FOrderController.java)|100%|:green_apple:|
|[OrderService.java](https://github.com/tkdwls4453/ecommerce-msa-server/blob/acb089df6cdedeca8419fcb32e851eb5b5fed113/order-service%2Fsrc%2Fmain%2Fjava%2Fcom%2Fmsa%2Forder%2Fapplication%2Fservice%2FOrderService.java)|100%|:green_apple:|
|[OrderItemEntity.java](https://github.com/tkdwls4453/ecommerce-msa-server/blob/acb089df6cdedeca8419fcb32e851eb5b5fed113/order-service%2Fsrc%2Fmain%2Fjava%2Fcom%2Fmsa%2Forder%2Fadapter%2Fout%2Fpersistence%2FOrderItemEntity.java)|100%|:green_apple:|
|[OrderCommandAdapter.java](https://github.com/tkdwls4453/ecommerce-msa-server/blob/acb089df6cdedeca8419fcb32e851eb5b5fed113/order-service%2Fsrc%2Fmain%2Fjava%2Fcom%2Fmsa%2Forder%2Fadapter%2Fout%2Fpersistence%2FOrderCommandAdapter.java)|100%|:green_apple:|
|[OrderEntity.java](https://github.com/tkdwls4453/ecommerce-msa-server/blob/acb089df6cdedeca8419fcb32e851eb5b5fed113/order-service%2Fsrc%2Fmain%2Fjava%2Fcom%2Fmsa%2Forder%2Fadapter%2Fout%2Fpersistence%2FOrderEntity.java)|100%|:green_apple:|
|[CreatedNewOrderResponse.java](https://github.com/tkdwls4453/ecommerce-msa-server/blob/acb089df6cdedeca8419fcb32e851eb5b5fed113/order-service%2Fsrc%2Fmain%2Fjava%2Fcom%2Fmsa%2Forder%2Fadapter%2Fin%2Fweb%2Fdto%2FCreatedNewOrderResponse.java)|100%|:green_apple:|
|[CreateNewOrderRequest.java](https://github.com/tkdwls4453/ecommerce-msa-server/blob/acb089df6cdedeca8419fcb32e851eb5b5fed113/order-service%2Fsrc%2Fmain%2Fjava%2Fcom%2Fmsa%2Forder%2Fadapter%2Fin%2Fweb%2Fdto%2FCreateNewOrderRequest.java)|100%|:green_apple:|
|[AppliedCoupon.java](https://github.com/tkdwls4453/ecommerce-msa-server/blob/acb089df6cdedeca8419fcb32e851eb5b5fed113/order-service%2Fsrc%2Fmain%2Fjava%2Fcom%2Fmsa%2Forder%2Fdomain%2Fvo%2FAppliedCoupon.java)|100%|:green_apple:|
|[OrderItem.java](https://github.com/tkdwls4453/ecommerce-msa-server/blob/acb089df6cdedeca8419fcb32e851eb5b5fed113/order-service%2Fsrc%2Fmain%2Fjava%2Fcom%2Fmsa%2Forder%2Fdomain%2Fvo%2FOrderItem.java)|100%|:green_apple:|
|[ShippingInfo.java](https://github.com/tkdwls4453/ecommerce-msa-server/blob/acb089df6cdedeca8419fcb32e851eb5b5fed113/order-service%2Fsrc%2Fmain%2Fjava%2Fcom%2Fmsa%2Forder%2Fdomain%2Fvo%2FShippingInfo.java)|100%|:green_apple:|
|[InsufficientStockException.java](https://github.com/tkdwls4453/ecommerce-msa-server/blob/acb089df6cdedeca8419fcb32e851eb5b5fed113/order-service%2Fsrc%2Fmain%2Fjava%2Fcom%2Fmsa%2Forder%2Fexception%2FInsufficientStockException.java)|0%|:x:|
|[NoOrderItemException.java](https://github.com/tkdwls4453/ecommerce-msa-server/blob/acb089df6cdedeca8419fcb32e851eb5b5fed113/order-service%2Fsrc%2Fmain%2Fjava%2Fcom%2Fmsa%2Forder%2Fexception%2FNoOrderItemException.java)|0%|:x:|
|[InvalidCouponException.java](https://github.com/tkdwls4453/ecommerce-msa-server/blob/acb089df6cdedeca8419fcb32e851eb5b5fed113/order-service%2Fsrc%2Fmain%2Fjava%2Fcom%2Fmsa%2Forder%2Fexception%2FInvalidCouponException.java)|0%|:x:|
|[FakeApplyCouponAdapter.java](https://github.com/tkdwls4453/ecommerce-msa-server/blob/acb089df6cdedeca8419fcb32e851eb5b5fed113/order-service%2Fsrc%2Fmain%2Fjava%2Fcom%2Fmsa%2Forder%2Fadapter%2Fout%2FFakeApplyCouponAdapter.java)|0%|:x:|
|[FakeDecreaseStockAdapter.java](https://github.com/tkdwls4453/ecommerce-msa-server/blob/acb089df6cdedeca8419fcb32e851eb5b5fed113/order-service%2Fsrc%2Fmain%2Fjava%2Fcom%2Fmsa%2Forder%2Fadapter%2Fout%2FFakeDecreaseStockAdapter.java)|0%|:x:|

Yundae203 commented on order-service/src/main/java/com/msa/order/domain/Order.java:
```diff
@@ -0,0 +1,102 @@
+    public static Order init(Long customerId, CreateNewOrderCommand command, LocalDateTime orderTime, UUID orderCode) {
+        verifyTotalPrice(command.orderLine(), command.coupon(), command.totalPrice());
+
+        return Order.builder()
+            .orderCode(orderCode)
+            .customerId(customerId)
+            .appliedCouponId(command.coupon() == null ? null : command.coupon().couponId())
+            .shippingInfo(command.shippingInfo())
+            .orderLine(command.orderLine())
+            .orderStatus(OrderStatus.ORDER_RECEIVED)
+            .totalPrice(command.totalPrice())
+            .orderTime(orderTime)
+            .build();
+    }
```
외부에서 3항 연산자로 체크하는 것보다,
CreateNewOrderCommand 내부에 getCouponId()를 따로 구현하는 것이 좋을 것 같습니다.

tkdwls4453 replied:
오 좋은 지적 감사합니다~

Yundae203 reviewed (approved):
전체적으로 깔끔하게 잘 작성된 것 같습니다.
고민의 흔적이 많이 묻어 나와서 좋습니다.

tkdwls4453 committed:
refactor: command 에서 쿠폰 아이디 가져오도록 수정 #40

쿠폰 아이디를 가져오는 함수를 command 에 추가하여 도메인에서는 코드를 깔끔하게 작성

tkdwls4453 committed:
refactor: Money 를 BigDecimal 로 관리 #40

정확한 계산을 보장하기 위해 안전한 BigDecimal 타입으로 변경

tkdwls4453 committed:
refactor: 할인 계산 책임을 Money 에서  처리 #40

할인율을 통해 할인값을 계산하는 책은 Money 객체에게 위임하였다.
테스트시 사용될 할인 적용 값을 매번 계산하지 않고 고정된 final 필드로 사용
